File: server/app/services/perspective_service.py
import json
from typing import Dict

from app.services.llama_service import LlamaService, llama_service


class PerspectiveService:
    """
    Service responsible for generating multi-perspective answers.
    Uses LLaMA with a strict JSON prompt and falls back to deterministic templates.
    """

    SYSTEM_PROMPT = """You are an ethical reasoning assistant.

Generate three concise perspectives from the provided material.

Return JSON only with this exact schema:
{
  "utilitarian": "...",
  "rights_based": "...",
  "care_ethics": "..."
}

Rules:
- utilitarian: focus on outcomes and collective wellbeing
- rights_based: focus on rights, duties, fairness
- care_ethics: focus on empathy, relationships, vulnerability
- Keep each field practical and distinct
- Do not add extra keys"""

    @classmethod
    async def generate_perspectives(
        cls,
        user_query: str,
        context: str,
        base_answer: str,
    ) -> Dict[str, str]:
        prompt = (
            f"USER QUERY:\n\"\"\"\n{(user_query or '').strip()}\n\"\"\"\n\n"
            f"CONTEXT:\n\"\"\"\n{(context or '').strip() or 'No prior context.'}\n\"\"\"\n\n"
            f"BASE ANSWER:\n\"\"\"\n{(base_answer or '').strip()}\n\"\"\""
        )

        try:
            raw = await llama_service.get_ai_response(prompt, system_prompt=cls.SYSTEM_PROMPT)
            parsed = cls._parse_json_payload(raw)
            return cls._normalize(parsed, base_answer)
        except (LlamaService.AIServiceError, ValueError, json.JSONDecodeError):
            return cls._build_local_perspectives(base_answer)

    @staticmethod
    def _parse_json_payload(raw: str) -> Dict[str, str]:
        text = str(raw or "").strip()
        if not text:
            raise ValueError("Empty LLaMA perspective response.")

        try:
            payload = json.loads(text)
        except json.JSONDecodeError:
            first_brace = text.find("{")
            last_brace = text.rfind("}")
            if first_brace == -1 or last_brace == -1 or last_brace <= first_brace:
                raise
            payload = json.loads(text[first_brace : last_brace + 1])

        if not isinstance(payload, dict):
            raise ValueError("Perspective response is not an object.")

        return payload

    @staticmethod
    def _normalize(payload: Dict[str, str], base_answer: str) -> Dict[str, str]:
        normalized = {
            "utilitarian": str(payload.get("utilitarian", "")).strip(),
            "rights_based": str(payload.get("rights_based", "")).strip(),
            "care_ethics": str(payload.get("care_ethics", "")).strip(),
        }

        if all(normalized.values()):
            return normalized

        fallback = PerspectiveService._build_local_perspectives(base_answer)
        for key, value in normalized.items():
            if not value:
                normalized[key] = fallback[key]

        return normalized

    @staticmethod
    def _build_local_perspectives(base_answer: str) -> Dict[str, str]:
        answer = " ".join(str(base_answer or "").split())
        if not answer:
            answer = "No answer was available to expand."

        return {
            "utilitarian": (
                "Utilitarian view: prefer the option that creates the most overall benefit "
                f"and the least overall harm. Base answer: {answer}"
            ),
            "rights_based": (
                "Rights-based view: ensure individual rights, consent, and fairness are preserved "
                f"while applying the recommendation. Base answer: {answer}"
            ),
            "care_ethics": (
                "Care ethics view: prioritize empathy, relationships, and the impact on vulnerable "
                f"people when applying this response. Base answer: {answer}"
            ),
        }


perspective_service = PerspectiveService()

File: server/services/perspectiveService.js
const HF_CHAT_COMPLETIONS_URL = "https://router.huggingface.co/v1/chat/completions";

const PERSPECTIVE_SYSTEM_PROMPT = [
  "You are an ethics reasoning assistant.",
  "Read the user question and context.",
  "Generate 3 perspectives: utilitarian, rights-based, and care ethics.",
  "",
  "Return STRICT JSON only with this exact schema:",
  "{",
  '  "utilitarian": "...",',
  '  "rights_based": "...",',
  '  "care_ethics": "..."',
  "}",
  "",
  "Rules:",
  "- utilitarian: focus on outcomes and overall well-being",
  "- rights_based: focus on rights, duties, and fairness",
  "- care_ethics: focus on relationships, vulnerability, and care",
  "- Keep each value concise and practical",
  "- Do not add extra keys",
].join("\n");

function asCleanString(value) {
  return String(value || "").trim();
}

function normalizeContext(context) {
  if (typeof context === "string") {
    const text = context.trim();
    return text || "No additional context provided.";
  }

  if (context == null) {
    return "No additional context provided.";
  }

  try {
    const serialized = JSON.stringify(context);
    return serialized && serialized !== "{}" ? serialized : "No additional context provided.";
  } catch {
    return "No additional context provided.";
  }
}

function normalizePerspectivePayload(payload) {
  return {
    utilitarian: asCleanString(payload?.utilitarian),
    rights_based: asCleanString(payload?.rights_based),
    care_ethics: asCleanString(payload?.care_ethics),
  };
}

function extractModelText(data) {
  const text = data?.choices?.[0]?.message?.content;

  if (!text || typeof text !== "string") {
    throw new Error("LLaMA returned an unexpected response format.");
  }

  return text.trim();
}

function parseJson(text) {
  try {
    return JSON.parse(text);
  } catch {
    const firstBrace = text.indexOf("{");
    const lastBrace = text.lastIndexOf("}");

    if (firstBrace === -1 || lastBrace === -1 || lastBrace <= firstBrace) {
      throw new Error("LLaMA did not return valid JSON.");
    }

    return JSON.parse(text.slice(firstBrace, lastBrace + 1));
  }
}

async function generatePerspectives(question, context) {
  if (typeof question !== "string" || !question.trim()) {
    throw new Error("generatePerspectives(question, context) requires a non-empty question string.");
  }

  const token = String(process.env.HF_TOKEN || "").trim();
  const model = String(process.env.HF_MODEL || "meta-llama/Llama-3.1-8B-Instruct").trim();

  if (!token) {
    throw new Error("Missing HF_TOKEN in environment variables.");
  }

  const userPrompt = [
    "USER QUESTION:",
    '"""',
    question.trim(),
    '"""',
    "",
    "CONTEXT:",
    '"""',
    normalizeContext(context),
    '"""',
  ].join("\n");

  const payload = {
    model,
    messages: [
      { role: "system", content: PERSPECTIVE_SYSTEM_PROMPT },
      { role: "user", content: userPrompt },
    ],
    max_tokens: 700,
    temperature: 0.3,
    top_p: 0.9,
    stream: false,
  };

  const response = await fetch(HF_CHAT_COMPLETIONS_URL, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`LLaMA API error ${response.status}: ${errorText}`);
  }

  const data = await response.json();
  const rawText = extractModelText(data);
  const parsed = parseJson(rawText);

  return normalizePerspectivePayload(parsed);
}

const perspectiveService = {
  generatePerspectives,
};

module.exports = {
  perspectiveService,
  generatePerspectives,
};
